docs/designs/high_performance_ps.md

```diff
@@ -0,0 +1,59 @@
+
+## Computation
+
+The gradients and parameters on PS are represented by tensors. And applying gradients to parameters, which is also called optimization, is acutally a math operation of tensors.
+
+### Tensor
+
+We have to support both dense tensor and sparse tensor. Besides, different element data types are also needed, such as int8/int32/float16/float32/float64. Int8 and float16 are used in training based quantization.
+
+Each tensor operator has to support different data types. C++ supports generics with template programming, while Go does not support generics directly.
+
+### Math library
+
+There are different kinds of optimizers, which need some tensor operations. There are many mature math libraries developed with C++. For example, [eigen](https://gitlab.com/libeigen/eigen) is used in TensorFlow and Paddle, [aten](https://github.com/pytorch/pytorch/tree/master/aten) is used in Pytorch. These math libraries provide abundant tensor operators and support both CPU and GPU. Besides, these math libraries could call some state-of-the-art blas libraries internally, such as MKL and cuBLAS. With these math libraries, the operators in optimizers could be implemented easily and efficiently.
+
+It seems that there are few math libraries in Go. [Gosl](https://github.com/cpmech/gosl) is no longer active, and [gonum](https://github.com/gonum/gonum) does not support MKL. Generally, the math library ecology of Go is far from competing to C++. And we also have some faint worry with the performance of math libraries in Go.
+
+## Scheduling
+
+In C++, we use thread based scheduling. Threads are scheduled by the operating system. Usually, we will implement a thread pool for computation, and another thread pool for IO. The parameter optimzation will be processed by the computation thread pool in parallel. In further, to reduce the overhead of context switching, we could bind a thread to a certain CPU core by setting CPU affinity to the thread. It will increase the cache hit rate of a CPU core.
+
+In Go, there is no concept of thread, we use goroutine instead. Goroutines are scheduled by Go runtime. Goroutine is not preemptive. There are four classes of events that occur in Go programs that allow the scheduler to make scheduling decisions. This does not mean it will always happen on one of these events. It means the scheduler gets the opportunity.
+
+- The use of the keyword `go`
+- Garbage collection
+- System calls
+- Synchronization and Orchestration
+
+Go supports concurrent programming well with first-class concepts, goroutine and channel.
+
+## Conclusion
+
+Considering the tradeoff between development efficiency and program peformance, we plan to put communication and scheduling parts in Go, and computation part in C++.
+
+[Cgo](https://golang.org/cmd/cgo/) enables the creation of Go packages that call C code. And the overhead of cgo is slight. The optimization operators will be implemented in C++, wrappered with C interface, and exposed to Go.
```

Comment:
Just FYI, the overhead of cgo cannot ignore in some case. There're many articles complain this problem:

- [The Cost and Complexity of Cgo](https://www.cockroachlabs.com/blog/the-cost-and-complexity-of-cgo): blog post from [CockroachDB](https://www.cockroachlabs.com) team
- [cgo is not Go](https://dave.cheney.net/2016/01/18/cgo-is-not-go): blog post from Dave Cheney (member of Go language)
- [Introducing Badger: A fast key-value store written purely in Go](https://blog.dgraph.io/post/badger/#cgo-the-necessary-evil): blog post from [Dgraph](https://dgraph.io) team, see the "Cgo: The necessary evil" section
- [go#19574](https://github.com/golang/go/issues/19574): cgo performance issue created by CockroackDB team

Reply:
@xiaogaozi Mant thanks for your kind reminder. 

There are mainly two kinds of overhead:

- Call overhead. I have made a benchmark, 1000-d dense tensor with Adam optimization, it seems that cgo gets about  8% more time than native C++. Please refer to https://github.com/QiJune/learning-notes/tree/master/test_codes/cgo.

- Scheduling overhead. In PS, cgo is used to call the parameter optimization function in C++. We will not launch hundreds of goroutines. It will be limited to be equal to the CPU core numbers.

So, let's move first. And I believe we could benefit from the speed-up of C++ and minimize the overhead.

Reply:
Very looking forward the Go language implementation of PS 🖖